lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
   }
 }
 
+static RecursiveCoroutine<bool> containsNonLocalCF(ASTNode *Node) {
+  switch (Node->getKind()) {
+  case ASTNode::NK_List: {
+    SequenceNode *Seq = llvm::cast<SequenceNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (ASTNode *N : Seq->nodes()) {
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(N);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Scs: {
+
+    // We do not want to move around a `ScsNode` for any reason
+    rc_return false;
+  } break;
+  case ASTNode::NK_If: {
+    IfNode *If = llvm::cast<IfNode>(Node);
+
+    bool HasNonLocalCF = false;
+    if (If->hasThen()) {
+      ASTNode *Then = If->getThen();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Then);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Else);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (auto &LabelCasePair : Switch->cases()) {
+      HasNonLocalCF = HasNonLocalCF
+                      or rc_recur containsNonLocalCF(LabelCasePair.second);
+    }
```

Comment:
Same as above, you can early exit at the first case that has non-local cf.

Reply:
Implemented the optimization [here](https://github.com/revng/revng-c/blob/f15b7b1a9506a37ed6d348f7dd6d880b1d93df16/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L560)